github.go

```diff
@@ -883,3 +883,67 @@ func (c *Client) ListReposBylanguage(owner string, lang string) ([]*github.Repos
 
 	return allRepos, nil
 }
+
+func (c *Client) ListReposOnlyBylanguage(lang string) ([]*github.Repository, error) {
+
+	query := Sf("language:%q", ToTitle(lang))
+
+	client := c.client
+
+	sortTypes := [3]string{"forks", "stars", "updated"}
+
+	for _, sortType := range sortTypes {
+		opt := &github.SearchOptions{
+			ListOptions: github.ListOptions{PerPage: 100},
+			Sort: sortType
+		}
+		// get all pages of results
+		var allRepos []*github.Repository
+		for {
+			var repos *github.RepositoriesSearchResult
+			var resp *github.Response
+			errs := RetryExponentialBackoff(9999, time.Second, func() error {
+				var err error
+
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+				defer cancel()
+
+				repos, resp, err = client.Search.Repositories(ctx, query, opt)
+				if err != nil {
+					return fmt.Errorf("error while executing request: %s", err)
+				}
+				ResponseCallback(resp)
+				if handleRateLimitError(err, resp) {
+					return err
+				}
+
+				if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNotFound && resp.StatusCode != http.StatusNoContent {
+					// TODO: catch rate limit error, and wait
+					return fmt.Errorf(
+						"status code is: %v (%s)",
+						resp.StatusCode,
+						resp.Status,
+					)
+				}
+				// nil on 200 and 404
+				return nil
+			})
+			if errs != nil && len(errs) > 0 {
+				return nil, errors.New(FormatErrorArray("", errs))
+			}
+			if resp.StatusCode == http.StatusNotFound {
+				// TODO: catch rate limit error, and wait
+				return nil, ErrNotFound
+			}
+
+			for repIndex := range repos.Repositories {
+				allRepos = append(allRepos, &repos.Repositories[repIndex])
+			}
+			if resp.NextPage == 0 {
+				break
+			}
+			opt.Page = resp.NextPage
+		}
+	}
+	return allRepos, nil
+}
```

Comment:
```suggestion
type ListReposOnlyBylanguageOpts struct {
	Language     string
	ExcludeForks bool
	MinStars     int
	Limit        int
}

// Validate validates ListReposOnlyBylanguageOpts.
func (opts *ListReposOnlyBylanguageOpts) Validate() error {
	if opts == nil {
		return errors.New("opts is nil.")
	}
	if opts.Language == "" {
		return errors.New("Language not specified.")
	}
	return nil
}

func (c *Client) ListReposOnlyBylanguage(opts *ListReposOnlyBylanguageOpts) ([]*github.Repository, error) {
	if err := opts.Validate(); err != nil {
		return nil, err
	}

	queryFragments := make([]string, 0)
	queryFragments = append(queryFragments, Sf("language:%q", ToTitle(opts.Language)))
	if opts.ExcludeForks {
		queryFragments = append(queryFragments, "fork:false")
	}

	client := c.client

	opt := &github.SearchOptions{
		Sort:        "stars", // Sort by stargazer count.
		ListOptions: github.ListOptions{PerPage: 100},
	}
	storeIndex := hashsearch.New()

	var (
		latestStarCount int
		useStarBound    bool
		starLowerBound  int
	)

	// get all pages of results
	var allRepos []*github.Repository
GetterLoop:
	for {
		var repos *github.RepositoriesSearchResult
		var resp *github.Response
		errs := RetryExponentialBackoff(9999, time.Second, func() error {
			var err error

			ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
			defer cancel()

			query := strings.Join(queryFragments, " ")
			if useStarBound {
				withBound := append(queryFragments, Sf("stars:<=%v", starLowerBound))
				query = strings.Join(withBound, " ")
			}

			repos, resp, err = client.Search.Repositories(ctx, query, opt)
			if err != nil {
				return fmt.Errorf("error while executing request: %s", err)
			}
			ResponseCallback(resp)
			if handleRateLimitError(err, resp) {
				return err
			}

			if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNotFound && resp.StatusCode != http.StatusNoContent {
				// TODO: catch rate limit error, and wait
				return fmt.Errorf(
					"status code is: %v (%s)",
					resp.StatusCode,
					resp.Status,
				)
			}
			// nil on 200 and 404
			return nil
		})
		if errs != nil && len(errs) > 0 {
			return nil, errors.New(FormatErrorArray("", errs))
		}
		if resp.StatusCode == http.StatusNotFound {
			// TODO: catch rate limit error, and wait
			return nil, ErrNotFound
		}
		for repIndex := range repos.Repositories {
			repo := &repos.Repositories[repIndex]

			if repo.GetStargazersCount() < opts.MinStars {
				break GetterLoop
			}
			id := repo.GetFullName()

			if !storeIndex.Has(id) {
				latestStarCount = repo.GetStargazersCount()

				allRepos = append(allRepos, repo)
				storeIndex.OrderedAppend(id)

				if opts.Limit > 0 && len(allRepos) >= opts.Limit {
					break GetterLoop
				}
			}
		}
		if resp.NextPage == 0 {
			// If we finished all the pages (10 x 100 results = 1K repos),
			// but the starLowerBound did not get lower,
			// that means there's more than 1K repos with that specific star count,
			// but we can't get more than 1K;
			// we can only retrieve the first 1K of repos with any specific star count.

			useStarBound = true
			if starLowerBound == latestStarCount {
				// For any given starLowerBound, we can only retrieve the first 1K repos.
				// For this particular starLowerBound,
				// there are more than 1K repos.

				// Let's decrement starLowerBound by one.
				// This will skip any repo with that specific star count beyond the initial 1K repos that we already got.
				latestStarCount--
			}
			starLowerBound = latestStarCount
			opt.Page = 1 // Restart
			continue GetterLoop
		}
		opt.Page = resp.NextPage
	}

	return allRepos, nil
}
```

Reply:
P.S. `hashsearch` is from `github.com/gagliardetto/hashsearch`

Reply:
Example:

With `language=GO`, there are about 26K repos before we reach the 18-star-repos.

Reply:
That's so cool! Thanks for the optimization (well, you actually rewrote it 👏), I wish I had more experience coding in Go to have been able to do that.

Btw, I have thought of taking repos also from a specific search (that would be awesome for testing specific queries), what's your view on that? (I'm not sure if google/go-github already supports it)

Thanks again! ^^

Reply:
> Btw, I have thought of taking repos also from a specific search (that would be awesome for testing specific queries), what's your view on that?

I personally haven't needed that (up till now I've always started ether from a list of GH repos, or GH accounts), but I see the value in having a free-form repo search integrated in `lgtm-cli`. Is that something you've found yourself needing? What was the particular use-case?

> (I'm not sure if google/go-github already supports it)

It would look something like this, but with a user-provided query: https://github.com/gagliardetto/gh-client/blob/9e34c265eb084fad619853228fbc13189eda575b/github.go#L827-L885

(BTW, the above `ListReposBylanguage` could be simplified by calling that new search function).

Remember that results would still be limited to 1K (and sorting by stars would disrupt the "Sort by best match" that github gives us, which has its own importance).

Reply:
> I personally haven't needed that (up till now I've always started ether from a list of GH repos, or GH accounts), but I see the value in having a free-form repo search integrated in `lgtm-cli`. Is that something you've found yourself needing? What was the particular use-case?

I'm testing queries against specific libraries, so searching for "import specific library" or "from lib import Module" would also be useful to improve some queries.

> It would look something like this, but with a user-provided query:
> 
> https://github.com/gagliardetto/gh-client/blob/9e34c265eb084fad619853228fbc13189eda575b/github.go#L827-L885

👏 

> Remember that results would still be limited to 1K (and sorting by stars would disrupt the "Sort by best match" that github gives us, which has its own importance).

Well, a thousand repos using an specific library the query is looking for is more powerful than having 50k repos of that language (from my point of view), so I thinks it's enough.

Reply:
> I'm testing queries against specific libraries, so searching for "import specific library" or "from lib import Module" would also be useful to improve some queries.

:exploding_head: You mean getting a list of repos based on a search inside the repo contents? That's an awesome idea! I see a lot of value in that. Haven't thought about that.

Reply:
But to do that, we'll need another function, and some light magic:

- Let's keep `SearchRepos`, which uses `client.Search.Repositories` (which only searches inside the name and description and meta of the repo), **BUT** also add a `SearchCode` which uses `client.Search.Code`, which will allow to make queries that return code matches and the repos they belong to.
- When `SearchCode` will return, you can get a list of repos from the results.

Reply:
So in the end we can have:

- `lgtm follow-by-search-repo "cloud"`
- `lgtm follow-by-search-code "from flask import Flask"`

Reply:
Oops, missed the last messages.

> 🤯 You mean getting a list of repos based on a search inside the repo contents? That's an awesome idea! I see a lot of value in that (and use cases for my future projects). Haven't thought about that.

Yeah exactly, the scope would be much focused on what you are querying.

> So in the end we can have:
> 
> * `lgtm follow-by-search-repo "cloud"`
> * `lgtm follow-by-search-code "from flask import Flask"`

That's perfect! ^^